### PR TITLE
Remove Clearance from password reset controller

### PR DESCRIPTION
**Pull request by martinemde**
The Clearance password reset controller behavior was mostly skipped or overridden. This completes the process and makes all the behavior clearly expressed in the passwords controller.

On moving functionality over, a few thing were changed:
- Sometimes the failure message on bad url would vary. Now it is consistent (and reworded to indicate clearly what should happen next).
- Added missing translations from clearance translation file.
- Correct a case where instead of redirecting to the root path on failure, it would render the passwords/new template on a request for edit. This may "make sense" in some ways, but it's not the behavior you would expect by reading the controller, but was hidden in clearance.
- Don't use the session[:password_reset_token] at all. You must finish the password reset flow continuously using the forms we render. This also prevents a user from abandoning a reset password flow at, e.g., the OTP challenge from persisting the state in their session that would allow continuing the password reset.

Additional thing I did not update that I think would be an improvement:
- [ ] It's weird that we render passwords/create.html.erb after submitting the passwords/new email form.
       It places you on a nearly empty page with only a message that could be in flash, and the page has no GET equivalent. Reloading the page will trigger the "are you sure you want to resubmit this form?". Overall this doesn't seem like a great design.
- [ ] I think it would make more sense for the password reset link to take you through a email token based login process, then redirect to password edit. That would set us up to allow password changes on verify while logged in without the email loop (technically this is possible now if you submit the `update` action directly after verifying using another action that request verification.)
- [ ] Instead of redirecting to root on token not found, I think it should redirect to `new_password_path` because that's what we're suggesting that they do to fix the problem.
- [ ] If you enter the wrong totp code, you will get the form re-rendered but it loses the token from the form action, so you can't submit it again even though it rendered. **I chose not to fix this because it would make it much easier to retry totp codes.**

### CODE REVIEW

File: app/controllers/passwords_controller.rb [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 
-  # By default, clearance expects the token to be submitted with the password update.
-  # We already invalidated the token when the user became verified by token(+mfa).
-  skip_before_action :ensure_existing_user, only: %i[update]
-  # Instead of the token, we now require the user to have been verified recently.
+  # Instead of the token for update, we require the user to have been verified recently.
   verify_session_before only: %i[update]
 
+  def new
+    render template: "passwords/new"
```
Comment (segiddins) [inline app/controllers/passwords_controller.rb:13]: I don't think an explicit `render` is needed here

File: app/controllers/passwords_controller.rb [hunk 2]
```diff
@@ -27,36 +28,49 @@ def edit
     end
   end
 
+  def create
+    email = params.dig(:password, :email)
+
+    if email.blank?
+      flash.now[:alert] = t(".failure_on_missing_email")
+      render template: "passwords/new", status: :unprocessable_entity
```
Comment (segiddins) [inline app/controllers/passwords_controller.rb:36]: move into a before_action

File: app/controllers/passwords_controller.rb [hunk 3]
```diff
@@ -27,36 +28,49 @@ def edit
...
-      current_user.api_keys.expire_all! if reset_params[:reset_api_keys] == "true"
-      redirect_to url_after_update
+    if current_user.update_password params.dig(:password_reset, :password)
+      current_user.reset_api_key! if reset_params[:reset_api_key] == "true" # singular
+      current_user.api_keys.expire_all! if reset_params[:reset_api_keys] == "true" # plural
+      redirect_to dashboard_path
       session[:password_reset_token] = nil
     else
-      flash_failure_after_update
+      # TODO: This should use the active record errors.
```
Comment (segiddins) [inline app/controllers/passwords_controller.rb:57]: TODO

File: app/controllers/passwords_controller.rb [hunk 4]
```diff
@@ -27,36 +28,49 @@ def edit
...
-    if otp_edit_conditions_met?
-      # When the user identified by the email token submits adequate totp, they are logged in
-      verified_sign_in
-      render template: "passwords/edit"
-    elsif !session_active?
-      login_failure(t("multifactor_auths.session_expired"))
-    else
-      login_failure(t("multifactor_auths.incorrect_otp"))
-    end
+    return login_failure(t("multifactor_auths.session_expired")) unless session_active?
```
Comment (segiddins) [inline app/controllers/passwords_controller.rb:64]: before action filter

File: app/controllers/passwords_controller.rb [hunk 5]
```diff
@@ -27,36 +28,49 @@ def edit
...
+    render template: "passwords/edit"
   end
 
   def webauthn_edit
-    unless session_active?
-      login_failure(t("multifactor_auths.session_expired"))
-      return
-    end
-
+    return login_failure(t("multifactor_auths.session_expired")) unless session_active?
```
Comment (segiddins) [inline app/controllers/passwords_controller.rb:73]: same as above, filter

File: app/controllers/passwords_controller.rb (hunk 1 above)
Comment (martinemde) [reply app/controllers/passwords_controller.rb:13]: Good point, copied this from clearance.

File: app/controllers/concerns/mfa_expiry_methods.rb [hunk 6]
```diff
@@ -12,7 +12,9 @@ def delete_mfa_expiry_session
 
     def session_active?
       return false if session[:mfa_expires_at].nil?
-      session[:mfa_expires_at] > Time.current
+      return true if session[:mfa_expires_at] > Time.current
+      delete_mfa_expiry_session
+      false
```
Comment (martinemde) [inline app/controllers/concerns/mfa_expiry_methods.rb:17]: This saves us from some ugliness when the before_filter prevents rendering on an expired session, preventing the after action form running that would clear `mfa_expires_at`.

File: app/controllers/passwords_controller.rb [hunk 7]
```diff
@@ -23,45 +27,41 @@ def edit
...
+
+  def create
+    user = User.find_by_normalized_email(@email)
+
+    if user
+      user.forgot_password!
+      ::PasswordMailer.change_password(user).deliver_later
     end
+
+    render :create, status: :accepted
```
Comment (martinemde) [inline app/controllers/passwords_controller.rb:42]: This lands you on a page that 404s if you hit enter on the url bar, and asks about form resubmit if you reload. I think it'd be better off going back to the signin page with a flash message set. What do you think?

File: app/controllers/concerns/mfa_expiry_methods.rb (hunk 6 above)
Comment (segiddins) [reply app/controllers/concerns/mfa_expiry_methods.rb:17]: please add a comment to that effect

**Review by segiddins — APPROVED** [review]

**Comment by martinemde** [comment]
@segiddins Thanks! I'll take notes for the other things I wanted to fix and follow up in another PR.

**Pull request merged**